### PR TITLE
Replace file_tester.py with tmpdir and tmpdir_factory

### LINKED ISSUE #2015: Delete cirq/testing/file_tester.py; use tmpdir and tmp_path pytest fixtures instead
http://doc.pytest.org/en/latest/tmpdir.html

### PR DESCRIPTION
**Pull request by andbe91**
Fixes #2015

### CODE REVIEW

File: dev_tools/bash_scripts_test.py [hunk 1]
```diff
@@ -24,13 +23,22 @@ def only_on_posix(func):
...
+        setup: str = '',
+) -> shell_tools.CommandOutput:
     """Invokes the given script within a temporary test environment."""
 
     with open(script_file) as f:
         script_lines = f.readlines()
 
+    # Create a unique temporary directory
+    dir_path = tmpdir_factory.mktemp('tmp', numbered=True)
+    file_path = f'{dir_path}/test-script'
```
Comment (cduck) [inline dev_tools/bash_scripts_test.py]: Don't hard code separator, file extension.

File: dev_tools/bash_scripts_test.py [hunk 2]
```diff
@@ -24,13 +23,22 @@ def only_on_posix(func):
     return func
 
 
-def run(*, script_file: str, arg: str ='', setup: str = ''
-        ) -> shell_tools.CommandOutput:
+def run(
+        *,
+        script_file: str,
+        tmpdir_factory,
```
Comment (cduck) [inline dev_tools/bash_scripts_test.py]: Is there a known type for this?

File: dev_tools/bash_scripts_test.py [hunk 3]
```diff
@@ -238,19 +263,24 @@ def test_pytest_and_incremental_coverage_branch_selection():
...
+                 tmpdir_factory=tmpdir_factory,
                  arg='HEAD~999999')
     assert result.exit_code == 1
     assert result.out == ''
     assert "No revision 'HEAD~999999'." in result.err
 
-    result = run(script_file='check/pytest-and-incremental-coverage')
+    result = run(
+        script_file='check/pytest-and-incremental-coverage',
+        tmpdir_factory=tmpdir_factory,
```
Comment (cduck) [inline dev_tools/bash_scripts_test.py]: Nit: Remove the trailing comma to make yapf format this the same as all the other `run` calls.

File: cirq/circuits/circuit_test.py [hunk 4]
```diff
@@ -2783,15 +2783,16 @@ def test_to_qasm():
 """.format(cirq.__version__))
 
 
-def test_save_qasm():
+def test_save_qasm(tmpdir):
+    file_path = f'{tmpdir}/qasm'
```
[the remark this answers is not in the document]
Comment (andbe91) [reply cirq/circuits/circuit_test.py]: Done,

File: cirq/circuits/qasm_output_test.py [hunk 5]
```diff
@@ -189,13 +189,13 @@ def test_version():
         _ = str(output)
 
 
-def test_save_to_file():
+def test_save_to_file(tmpdir):
+    file_path = f'{tmpdir}/qasm'
```
[the remark this answers is not in the document]
Comment (andbe91) [reply cirq/circuits/qasm_output_test.py]: Done.

File: dev_tools/bash_scripts_test.py (hunk 1 above)
Comment (andbe91) [reply dev_tools/bash_scripts_test.py]: Are you referring to `tmp`, the name of the temporary directory? Do you want it as an argument to the function instead?

File: dev_tools/bash_scripts_test.py (hunk 2 above)
Comment (andbe91) [reply dev_tools/bash_scripts_test.py]: Added.

File: dev_tools/bash_scripts_test.py (hunk 3 above)
Comment (andbe91) [reply dev_tools/bash_scripts_test.py]: Done.

File: dev_tools/bash_scripts_test.py (hunk 1 above)
Comment (cduck) [reply dev_tools/bash_scripts_test.py]: I just mean use `os.path.join` and call the file something like "test-script.sh".

Comment (andbe91) [reply dev_tools/bash_scripts_test.py]: Ok, done.

Comment (cduck) [reply dev_tools/bash_scripts_test.py]: I didn't realize that "test-script" was already named.  Either way is fine.

Comment (andbe91) [reply dev_tools/bash_scripts_test.py]: I agree that it makes sense to give the file its proper extension.

**Pull request merged**